File: server/circuitBreaker.mjs
// Minimal dependency-free circuit breaker for outbound calls (model API, Octane).
//
// Why this exists: companyChat.mjs already retries individual requests, but if an
// upstream stays down, every incoming request still pays the full retry budget
// before failing. A circuit breaker trips after N consecutive failures and
// fast-fails subsequent calls (returning a degradation instead of waiting out the
// timeout), then probes with one trial call after resetTimeoutMs.
//
// States: closed (normal, count failures) -> open (fast-fail) -> half-open (one
// trial). A success in half-open closes the breaker; a failure re-opens it.
//
// Failure accounting: thrown errors always count. A non-throwing result also
// counts as a failure when isFailure(null, result) is true (e.g. HTTP 5xx): the
// result is still returned to the caller (so the existing response handling and
// diagnostics run unchanged), but the breaker tally increments and may trip on
// accumulated upstream errors.

export class CircuitOpenError extends Error {
  constructor(message, { service } = {}) {
    super(message);
    this.name = "CircuitOpenError";
    this.service = service || "unknown";
    this.circuitOpen = true;
  }
}

export function createCircuitBreaker({
  service = "unknown",
  failureThreshold = 5,
  resetTimeoutMs = 30_000,
  halfOpenMaxCalls = 1,
  // Decide whether a given call outcome counts as a breaker failure. Thrown
  // errors count by default; callers can also flag a non-throwing result (e.g.
  // an HTTP 5xx response) so the breaker trips on accumulated upstream errors
  // without the caller having to throw.
  isFailure = (error, result) => Boolean(error),
} = {}) {
  let state = "closed";
  let failureCount = 0;
  let lastFailureAt = 0;
  let halfOpenInflight = 0;

  function maybeHalfOpen() {
    if (state === "open" && Date.now() - lastFailureAt >= resetTimeoutMs) {
      state = "half-open";
      halfOpenInflight = 0;
    }
  }

  function recordFailure() {
    failureCount += 1;
    lastFailureAt = Date.now();
    if (failureCount >= failureThreshold) state = "open";
  }

  function recordSuccess() {
    state = "closed";
    failureCount = 0;
    halfOpenInflight = 0;
  }

  async function call(fn) {
    maybeHalfOpen();
    if (state === "open") {
      throw new CircuitOpenError(
        `${service} circuit open (failures=${failureCount})`,
        { service },
      );
    }
    if (state === "half-open" && halfOpenInflight >= halfOpenMaxCalls) {
      throw new CircuitOpenError(
        `${service} circuit half-open throttled`,
        { service },
      );
    }
    if (state === "half-open") halfOpenInflight += 1;

    let result;
    try {
      result = await fn();
    } catch (err) {
      // Thrown call: always a failure. In half-open this reopens immediately.
      state = "open";
      lastFailureAt = Date.now();
      halfOpenInflight = 0;
      throw err;
    }

    const failed = Boolean(isFailure(null, result));
    if (state === "half-open") {
      // Trial result decides the breaker but is always returned to the caller.
      if (failed) {
        state = "open";
        lastFailureAt = Date.now();
      } else {
        recordSuccess();
      }
      halfOpenInflight = 0;
    } else if (failed) {
      recordFailure();
    } else {
      recordSuccess();
    }
    return result;
  }

  function getState() {
    return { service, state, failureCount, lastFailureAt };
  }

  return { call, getState };
}

File: backend/analytics/circuit_breaker.py
"""Minimal dependency-free circuit breaker for outbound calls (Octane, etc.).

Symmetric with server/circuitBreaker.mjs on the Node side. Used by the Octane
client so a sustained upstream outage fast-fails ingest instead of every caller
paying the full timeout budget. States: closed -> open -> half-open.

Failure accounting: thrown errors count when ``is_failure(err, None)`` is true;
non-throwing results count when ``is_failure(None, result)`` is true (e.g. an HTTP
5xx response). The result is always returned to the caller unchanged so the
existing raise_for_status() diagnostics keep working.
"""
from __future__ import annotations

import time
from typing import Any, Callable, Optional


class CircuitOpenError(Exception):
    """Raised when the breaker is open and a call is fast-failed."""

    def __init__(self, message: str, *, service: str = "unknown") -> None:
        super().__init__(message)
        self.service = service
        self.circuit_open = True


def _status_from(obj: Any) -> Optional[int]:
    status = getattr(obj, "status_code", None)
    if isinstance(status, int):
        return status
    response = getattr(obj, "response", None)
    status = getattr(response, "status_code", None)
    return status if isinstance(status, int) else None


def is_http_failure(error: Optional[BaseException], result: Any) -> bool:
    """Default is_failure predicate for HTTP clients.

    Counts transport errors (thrown) and 5xx/408/429 responses. A 4xx response or
    a 4xx HTTPError (raised by raise_for_status) does NOT count: those are request
    errors, not evidence the upstream is down.
    """
    if error is not None:
        status = _status_from(error)
        if status is not None and 400 <= status < 500 and status not in (408, 429):
            return False
        return True
    status = _status_from(result)
    if status is None:
        return False
    return status == 408 or status == 429 or status >= 500


class CircuitBreaker:
    def __init__(
        self,
        *,
        service: str = "unknown",
        failure_threshold: int = 5,
        reset_timeout_s: float = 30.0,
        half_open_max_calls: int = 1,
        is_failure: Optional[Callable[[Optional[BaseException], Any], bool]] = None,
    ) -> None:
        self.service = service
        self.failure_threshold = max(1, int(failure_threshold))
        self.reset_timeout_s = max(0.0, float(reset_timeout_s))
        self.half_open_max_calls = max(1, int(half_open_max_calls))
        self._is_failure = is_failure or (lambda error, result: error is not None)
        self._state = "closed"
        self._failure_count = 0
        self._last_failure_at = 0.0
        self._half_open_inflight = 0

    def _maybe_half_open(self) -> None:
        if self._state == "open" and time.monotonic() - self._last_failure_at >= self.reset_timeout_s:
            self._state = "half-open"
            self._half_open_inflight = 0

    def _record_failure(self) -> None:
        self._failure_count += 1
        self._last_failure_at = time.monotonic()
        if self._failure_count >= self.failure_threshold:
            self._state = "open"

    def _record_success(self) -> None:
        self._state = "closed"
        self._failure_count = 0
        self._half_open_inflight = 0

    def call(self, fn: Callable[[], Any]) -> Any:
        self._maybe_half_open()
        if self._state == "open":
            raise CircuitOpenError(
                f"{self.service} circuit open (failures={self._failure_count})",
                service=self.service,
            )
        if self._state == "half-open" and self._half_open_inflight >= self.half_open_max_calls:
            raise CircuitOpenError(
                f"{self.service} circuit half-open throttled",
                service=self.service,
            )
        if self._state == "half-open":
            self._half_open_inflight += 1
        was_half_open = self._state == "half-open"
        try:
            result = fn()
        except Exception as err:
            failed = bool(self._is_failure(err, None))
            if failed:
                if was_half_open:
                    self._state = "open"
                    self._last_failure_at = time.monotonic()
                else:
                    self._record_failure()
            self._half_open_inflight = 0
            raise
        failed = bool(self._is_failure(None, result))
        if self._state == "half-open":
            if failed:
                self._state = "open"
                self._last_failure_at = time.monotonic()
            else:
                self._record_success()
            self._half_open_inflight = 0
        elif failed:
            self._record_failure()
        else:
            self._record_success()
        return result

    def get_state(self) -> dict[str, Any]:
        return {
            "service": self.service,
            "state": self._state,
            "failureCount": self._failure_count,
            "lastFailureAt": self._last_failure_at,
        }
